### DISCUSSION TITLE
Using lifespan event on `APIRoute()` instead in `FastAPI()`

### DISCUSSION THREAD

**Discussion by ammarsaf** [fastapi/fastapi#9664] [Questions] [answered]
### Privileged issue

- [X] I'm @tiangolo or he asked me directly to create an issue here.

### Issue Content

I have asked Mr @tiangolo on Twitter this question and he suggested me to ask in Github so it can benefit others and I am agree with him. 

I have tried to apply lifespan event on `APIrouter()` instead of `app` object, based on the FastAPI [tutorial](https://fastapi.tiangolo.com/advanced/events/?h=). But it seems that to showed the Keyword error on `"answer_to_everything"`. I have search on Google but I don't found any discussion related to this. 

I found that `APIRoute()` class has `lifespan` parameter too (same as `FastAPI()`), but might be require certain or additional configuration to make it happened. Any way that we can apply the lifespan on router?

I hope anyone can help to solve this issue. Thank you for creating an awesome API framework.

**Comment by Kludex** [comment, marked as answer]
> I have asked Mr @tiangolo on Twitter this question and he suggested me to ask in Github so it can benefit others and I am agree with him.

He probably meant to create a discussion, not an issue.

> I found that APIRoute() class has lifespan parameter too (same as FastAPI()), but might be require certain or additional configuration to make it happened. Any way that we can apply the lifespan on route?

The `APIRouter` is the one that accepts `lifespan`, not `APIRoute`.

> Any way that we can apply the lifespan on route?

For the moment, is not possible.

**Reply by ammarsaf** [reply]
Alright thanks @Kludex for the answer.

**Reply by stur86** [reply]
Adding it to `APIRouter` doesn't work either for me, though. Only `FastAPI` seems to properly execute the lifespan context methods. Is something missing?

**Reply by mastertilla** [reply]
Hi @Kludex, first of all thank you for all the work with FastAPI and Pydantic! I would like to point out that I don't think you really answer the question and hence it should not be marked as the right answer (as already pointed out by @joshwilson-dbx). I think the underlying problem is still there (be that an issue with FastAPI or Starlette).

I think this issue requires a proper answer, as it seems to be the meeting point for everyone encountering the issue with deprecated `on_event` issues. I think the community is very keen to at least get an update on this (answer is from more than a year ago), whether that is that the work is in progress or that the ability to run events on routers is not longer supported.

**Reply by Kludex** [reply]
- This was fixed in 0.112.2 on https://github.com/fastapi/fastapi/pull/9630

```py
from contextlib import asynccontextmanager
from fastapi import APIRouter, FastAPI

@asynccontextmanager
async def lifespan(app: FastAPI):
    print("Enter")
    yield
    print("Exit")

app = FastAPI(lifespan=lifespan)

@app.get("/")
def read_root():
    return {"Hello": "World"}

router = APIRouter(lifespan=lifespan)
app.include_router(router)
```

Note that `APIRoute` doesn't have the `lifespan` parameter, only `APIRouter` does.

**Reply by mastertilla** [reply]
Thank you for your quick response 🙌🏽

**Comment by noirbizarre** [comment]
Hi 👋🏼 

I was going to submit an issue, but I saw this discussion.
Currently `on_startup` and `on_shutdown` are deprecated in favor of `lifespan` on both `FastAPI` and `APIRouter` while `lifespan` are only supported on `FastAPI`.

Here is a quick gist showing `APIRouter` does not support `lifespan`: https://gist.github.com/noirbizarre/a6028f561adb5949c7f21471b9b02032

What is FastAPI official position on this ?
1. `APIRouter`-level events are not supported anymore
2. `APIRouter`-level lifespan support is missing and this is an issue that need a fix
3. `APIRouter`-level events are not deprecated and the warning is app warning is leaking on `APIRouter` and this need a fix

**Reply by mcjohnson1** [reply]
Also encountering this issue with `APIRouter`.

**Reply by naimetti** [reply]
Same here

**Reply by joshwilson-dbx** [reply]
@Kludex I might be misreading your answer, but it looks like this discussion was resolved by you just clearing up some confusion about the difference between `APIRoute` and `APIRouter`. From my reading it looks like you're saying the intent is for `APIRouter` to fully support the lifespan functionality.

I can also confirm that a lifespan asynccontextmanager function is not invoked when provided to a sub-`APIRouter`. Is this intentional?

The FastAPI docs _do_ mention that this is not supported for sub-applications, but perhaps that should be updated to include sub-routers, too. At least for the time being. For now, I think we'll have to fall back to the deprecated lifecycle events pattern or something like this:

```
def startup():
    logger.info("Starting sub router")


router = APIRouter(on_startup=[startup])

app = FastAPI()
app.include_router(router, '/myprefix')
```

Looking into the code, I don't see `FastAPI` doing anything with the app's lifespan argument other than handing it to its internal `APIRouter` instance. I think we could support the sub-router case (and perhaps the sub-application case). To my uneducated eyes, it seems like it would be a relatively easy thing to have the `FastAPI` app wrap it's own lifespan manager around any user-provided one and which would invoke any sub-routers' lifespan managers before invoking the user-provided app-level one (or maybe the user-provided app one is invoked first?). Would a PR be welcome?

**Reply by noirbizarre** [reply]
I think the fact that:
- the [`APIRouter` documentation](https://fastapi.tiangolo.com/reference/apirouter/) itself is saying explicitely that `on_startup` and `on_shutdown` are deprecated in favor or `lifespans`
- `DeprecationWarnings` are raised when you use `on_startup` and `on_shutdown`
- the `lifespan` parameter is silently ignored

are misleading and either:
- it should be fixed (I can try to submit a fix if a PR is welcome) if `lifespan` is expected to work on `APIRouter`
- documentation should be fixed and `DeprecationWarning` removed on `APIRouter` `on_startup` and `on_shutdown` parameters if support for lifespan is not intended

**Reply by nicocti** [reply]
True. Actually, a PR is already opened here : https://github.com/tiangolo/fastapi/pull/9630 but it seems that it should be fixed upward (in Starlette). Unfortunately, Starlette won't implement before 1.0 (if I got it right here: https://github.com/encode/starlette/pull/1988#issuecomment-1416843513). I'd also benefit from using lifespan on routers on my side.

**Reply by kamalkraj** [reply]
> Hi 👋🏼
> 
> I was going to submit an issue, but I saw this discussion. Currently `on_startup` and `on_shutdown` are deprecated in favor of `lifespan` on both `FastAPI` and `APIRouter` while `lifespan` are only supported on `FastAPI`.
> 
> Here is a quick gist showing `APIRouter` does not support `lifespan`: https://gist.github.com/noirbizarre/a6028f561adb5949c7f21471b9b02032
> 
> What is FastAPI official position on this ?
> 
> 1. `APIRouter`-level events are not supported anymore
> 2. `APIRouter`-level lifespan support is missing and this is an issue that need a fix
> 3. `APIRouter`-level events are not deprecated and the warning is app warning is leaking on `APIRouter` and this need a fix

```python
from contextlib import asynccontextmanager

from fastapi import APIRouter, FastAPI, Request


@asynccontextmanager
async def lifespan(app: FastAPI):
    # Load the ML model
    app.ml_model = lambda x: x * 42
    yield
    # Clean up the ML models and release the resources
    del app.ml_model


app = FastAPI(lifespan=lifespan)
router = APIRouter()


@router.get("/predict")
async def predict(request: Request, x: float):
    result = request.app.ml_model(x)
    return {"result": result}


app.include_router(router)
```

**Reply by fclesio** [reply]
Hit the same issue. 

My use case is that I create a single main app (FastAPI), with multiple routing along the some sub-module folder structure that I import later on. 

If I use the lifespan in the main as the @noirbizarre mentioned does not work for me as well.

**Comment by Blackwidow-sudo** [comment]
Just ran into the same problem.
Makes no sense to me that `APIRouter` has the kwarg for `lifespan` but doesn't actually do anything.

**Comment by Harrolee** [comment]
In the meantime, this [SO answer](https://stackoverflow.com/questions/76322463/how-to-initialise-a-global-object-or-variable-and-reuse-it-in-every-fastapi-endp/76322910#76322910) seems to be the internet's chosen workaround.

**Reply by mastertilla** [reply]
I've actually used this pattern to inject instances into dependency injection and it works really well 👍🏽
